packages/app/src/specs/generators/component/ReactComponentGeneratorStepOne.vue

```diff
@@ -0,0 +1,180 @@
+import { ReactComponentGeneratorStepOneDocument, ReactComponentGeneratorStepOne_GenerateSpecDocument } from '../../../generated/graphql'
+import StandardModalFooter from '@cy/components/StandardModalFooter.vue'
+import Button from '@cy/components/Button.vue'
+import PlusButtonIcon from '~icons/cy/add-large_x16.svg'
+import TestResultsIcon from '~icons/cy/test-results_x24.svg'
+import EmptyGenerator from '../EmptyGenerator.vue'
+import { posixify } from '../../../paths'
+
+const props = defineProps<{
+  title: string
+  gql: ComponentGeneratorStepOne_CodeGenGlobFragment
+}>()
+const { t } = useI18n()
+const emits = defineEmits<{
+  (event: 'update:title', value: string): void
+  (event: 'update:description', value: string): void
+  (event: 'restart'): void
+  (event: 'close'): void
+}>()
+const { title } = useVModels(props, emits)
+
+title.value = t('createSpec.component.importFromComponent.chooseAComponentHeader')
+
+gql`
+query ReactComponentGeneratorStepOne($glob: String!) {
+  currentProject {
+    id
+    codeGenCandidates(glob: $glob) {
+      id
+      fileName
+      fileExtension
+      absolute
+      relative
+      baseName
+    }
+    # Add the specs, so we can keep the list up to date with the cache
+    specs {
+      id
+      ...SpecNode_InlineSpecList
+    }
+  }
+}
+`
+
+gql`
+mutation ReactComponentGeneratorStepOne_generateSpec($codeGenCandidate: String!, $type: CodeGenType!, $componentName: String!, $isDefault: Boolean!) {
+  generateSpecFromSource(codeGenCandidate: $codeGenCandidate, type: $type, componentName: $componentName, isDefault: $isDefault) {
+    ...GeneratorSuccess
+    currentProject {
+      id
+      ...EmptyGenerator
+    }
+    generatedSpecResult {
+      ... on GeneratedSpecError {
+        fileName
+      }
+    }
+  }
+}`
+
+const generateSpecMutation = useMutation(ReactComponentGeneratorStepOne_GenerateSpecDocument)
+const extensionPattern = ref(props.gql.codeGenGlobs.component)
+
+const query = useQuery({
+  query: ReactComponentGeneratorStepOneDocument,
+  // @ts-ignore
+  variables: { glob: extensionPattern },
+})
+const allFiles = computed((): any => {
+  if (query.data.value?.currentProject?.codeGenCandidates) {
+    return query.data.value.currentProject?.codeGenCandidates
+  }
+
+  return []
+})
+const result = ref<GeneratorSuccessFileFragment | null>(null)
+const generatedSpecError = ref()
+const generateSpecFromSource = ref()
+
+whenever(result, () => {
+  title.value = t('createSpec.successPage.header')
+})
+
+whenever(generatedSpecError, () => {
+  title.value = t('createSpec.component.importEmptySpec.header')
+})
+
+const makeSpec = async ({ file, item }) => {
+  const { data } = await generateSpecMutation.executeMutation({
+    codeGenCandidate: file.absolute,
+    type: 'component',
+    componentName: item.exportName,
+    isDefault: item.isDefault,
+  })
+
+  generateSpecFromSource.value = data?.generateSpecFromSource
```

Comment:
Woah, this is quite hard to read. Should we have a field on the mutation response that tells us what we want to know, instead of having to infer it like this?

Reply:
yeah this is pretty complex. My understanding is that the type that this mutation returns is a union type, so it either returns a `ScaffoldedFile` response or a `GeneratedSpecError` response. In order to know which response we got, we need to check the `__typename` field like this.

There's probably a better way to refactor this... I wonder if we could make an issue to go back and refactor these types later?

Reply:
Refactor later means refactor never - happy to live with this for now, but I hope we can start to think about how to get a handle of GraphQL. Things are getting quite messy, I think the core problem is we should be handling this on the back-end and giving the client a more well defined response in the first place. It would eliminate the need for the conditionals in the frontend.